Adding backend code for editing poll question and answers

balajisubr opened the pull request:
:construction: WIP :construction: 
https://github.com/electology/approval_frame/issues/108

Faxn commented:
Nice. I looked over your code and I have a few comments, but I must preface it with a disclaimer that what I say is a mere suggestion with no weight, and you might want to wait for one of the maintainers to comment on you pull request before you actually change anything.
* You added `can_edit` to the poll model but don't seem to ever use it. I tested your version and it lets you edit questions that have votes already, which is something the issue referenced says should not be allowed. I think that it might be worth discussing if this requirement makes sense, but the inclusion of `can_edit` leads me to believe that this is an accidental omission instead of deliberate divergence from the spec.
* It would probably be best to take the logging statements out or at least change them to `INFO` level. The content they log isn't an error and doesn't need to show up at that priority.

rupalkhilari commented:
@Faxn thanks for your comments, very valid indeed ! @balajisubr did you mean to indicate that this PR is still work in progress ?

balajisubr commented:
@Faxn thanks for the suggestions! and yea @rupalkhilari this is indeed  a PR in progress as indicated by the construction symbol! :)  sorry if that wasn't self-explanatory!

balajisubr commented:
@rupalkhilari @Faxn Can you review the changes for this? There are two things that need to be fixed before this can be merged. 1) I have implemented poll.can_edit() but it looks like you can't use a method in django templates with parantheses. so I have poll.can_edit which returns incorrect results. I am not sure how to return the proper results 2) With the current code, it looks like the + button in the screenshot for adding choices has the action http://bit.ly/2riq4bv but when I change the value of this.numChoiceFields on line 3 to 41, I expect the text box opened by the + button to have a value of 'Choice 42' as placeholder text but it does not. It gets the value of 'Choice 5'. I am trying to reuse this action for the edit button action to have the right placeholder text but found this weird behavior. So am not sure what else is causing the value to be 5 instead of 42. Besides that, I have updated the logic for user subscriptions script. Let me know if there are any questions.

rupalkhilari commented:
Thanks @balajisubr ! I shall take a look at it today and get back to you about it.

rupalkhilari commented on approval_polls/static/approval_polls/create.js:
```diff
@@ -27,7 +29,8 @@ $(function () {
   $('[data-toggle=tooltip]').tooltip();
-  $('button#add-choice').click($.proxy(this.addChoiceField, this));
+  $('button#add-choice').click(addChoiceField(this.numChoiceFields));
+  $('button#add-choice-edit').click(addChoiceField(Number($('div#choices-count').text()))
```
I don't seem to be able to add a new option as it gives this error:
create.js:33 Uncaught SyntaxError: missing ) after argument list

balajisubr replied:
have fixed the syntax error

rupalkhilari commented on approval_polls/templates/approval_polls/edit.html:
```diff
@@ -16,111 +16,113 @@
+        <button type='submit' class='btn btn-success pull-right'>Save</button>
+      </form>
+    {% else %}
+      <h3>The poll cannot be edited!</h3>
```
It would be nice if you can explain to the user why the poll cannot be edited.

balajisubr replied:
have changed the message

rupalkhilari commented on approval_polls/models.py:
```diff
@@ -44,7 +46,30 @@ def voters(self):
 
-
+    def can_edit(self):
+        return self.total_ballots == 0
```
`self.total_ballots` is a method of Poll and hence is always returning False when compared to 0. You'd need to use `self.total_ballots()`

rupalkhilari commented:
1) You are right @balajisubr . You have implemented it as required, but it seems to be returning False each time since a method is being compared with an integer. See comments above.
2) hmm..I'm not sure why you are not seeing the placeholder text update. I tried changing line 3 in the code on the master branch and I see what is expected i.e. Choice 42 placeholder text and so on. Is this what you meant ?
<img width="903" alt="screen shot 2017-06-12 at 2 22 45 am" src="https://user-images.githubusercontent.com/7693347/27028334-620c68b4-4f19-11e7-9903-a2446eba2d4c.png"> 

Thanks for your work, let me know if you have any other questions!

balajisubr commented:
@rupalkhilari have made requested changes.  Let me know if there are any further questions. There is one minor bug I have found that already exists in master which I found when testing out adding choices. If you add 2 options example Choice 5 and Choice 6 and delete Choice 6, then add another one, the latest one also has a placeholder/id of 6. I am not an expert at making UI changes so not sure how to fix that one elegantly. This affects editing polls because if there are 2 new choices which have the same id, only one gets saved in the database.

Faxn commented:
I completely missed the WIP sign somehow, sorry. Is it normal to submit a pull request that is under construction and not ready to merge?

I made an issue(#132) for the bug you mentioned and I'm going to submit a PR for it right after this.

rupalkhilari commented:
Thanks @balajisubr, I shall review your changes and back to you today. @Faxn normally we do not submit a work in progress pull request, but you are welcome to in case it is more convenient to you or if you need some feedback.

rupalkhilari commented on approval_polls/models.py:
```diff
@@ -1,3 +1,5 @@
+import logging
```
Unused import here. Could you check and fix any style violations in the code you are changing by running the flake8 command as given in the README page (https://github.com/electology/approval_frame#coding-style) ?

rupalkhilari commented on approval_polls/static/approval_polls/create.js:
```diff
@@ -120,7 +126,10 @@ $(function () {
     console.log(e);
     var container = $(e.currentTarget).closest('.input-group');
     container.remove();
-    that.numChoiceFields--;
```
I believe @Faxn is working to iron out this issue of choice numbers, so would you like to introduce your changes after his are committed ?

rupalkhilari commented on approval_polls/templates/approval_polls/edit.html:
```diff
@@ -16,111 +16,113 @@
+        <button type='submit' class='btn btn-success pull-right'>Save</button>
+      </form>
+    {% else %}
+      <h3>The poll cannot be edited as it has ballots on it!</h3>
```
With this change the user is not able to edit anything about the poll even if it has a single ballot on it. I think we should allow the poll owner to change the other parameters such as poll close date and visibility on the home page. However we would not want to allow the text on the choice or poll question to be edited.

balajisubr commented:
Thanks for creating the bug @Faxn! I will make sure I don't submit WIP PRs as they are confusing. @rupalkhilari You are right. I will wait till @Faxn 's changes are merged as I can use them for mine. And I will implement the remaining changes sometime in the next couple changes. Thanks for the feedback!

balajisubr commented:
@rupalkhilari Can you review this once again and merge if ok?

Faxn commented on approval_polls/templates/approval_polls/edit.html:
```diff
@@ -16,111 +16,117 @@
+        {% for choice in choices %}
+        <div class='form-group'>
+          <div class='input-group' id='div-choice{{ choice.pk }}'>
+            <input class='form-control' type='text' maxlength='200' name='choice{{ choice.pk }}' placeholder='Choice' value='{{choice.choice_text}}' {% if not can_edit_poll %}disabled{% endif %}>
```
Choice.pk is based on the database id of the choice. It might be 5. If it's 5 and a user edits the poll and adds another choice there will be two `choice5`s this will cause problems. The fix to this probably happens in create.js, you need some way to start the ids higher than the highest existing id.

Faxn commented on approval_polls/models.py:
```diff
@@ -44,6 +44,33 @@ def voters(self):
+    def can_edit(self):
+        return self.total_ballots() == 0
+
+    def add_choices(self, ids, choice_data):
```
I am a little hesitant about these methods that encourage passing `request.POST` right to the model. It seems like It might violate some kind of separation of concerns best practices or something. If this doesn't bother anyone else I'll say no more.

balajisubr commented:
@Faxn Thanks for the inputs! They are all valid points. I have implemented them. 
@rupalkhilari I realized that I still need to validate inputs for blank choices for this PR to be complete. If a user enters a blank choice, should the page be refreshed and the blank choice be  shown in red?

balajisubr commented on approval_polls/static/approval_polls/create.js:
```diff
@@ -1,33 +1,41 @@
+$(function() {
+  var that = this; 
+  this.numChoiceFields = 4
+  this.currentNum = undefined;
```
@Faxn What do you mean by data path? I added the currentNum fields to get the highest ID of the choice. With other changes I have now implemented, hope the use of currentNum makes sense.

Faxn replied:
"Data path" is perhaps a bad term. Specifically I was referring to how you broke up `this.numChoiceFields` into `this.numChoiceFields`, `this.currentNum` and the parameter `numChoiceFields`. 

Looking at it now I sort of see what you were doing. I think `this.NumCoiceFields` can be removed safely. It just sits there holding it's `4` and is never read or written to that I can see. 

`#choices-count` is named counter intuitively, Something like `#LastId` would be better. It might be a good idea to rename the javascript variable used to determine the next id(`this.currentNum`) into `this.LastId`, since Id numbers no longer have much to do with the number of choices. It's optional though.

rupalkhilari commented:
@balajisubr , the create page skips all blank inputs by default, so I think we should keep that behavior consistent in the edit page too. What do you think ?

rupalkhilari commented:
@balajisubr, I noticed another issue, it is not possible to edit the poll closing date and other date settings if the poll has a ballot on it. Although it is enabled, it does not save the changes.

rupalkhilari commented:
The changes look good. Thanks a lot for your work, @balajisubr !
